File: python/API/US003PostCourses.py
import requests
import json

# --- Konfigürasyon ---
BASE_URL = "https://qa.instulearn.com"
# Token süresi dolduysa güncellemeyi unutmayın
TOKEN = "15707|XCnyE8Ov4ZyvaMbZIbMfzOpRUbiyA1eUVPyUhXOB"

HEADERS = {
    "Accept": "application/json",
    "x-api-key": "1234",
    "Authorization": f"Bearer {TOKEN}"
}

def test_post_courses_tc01():
    """US003 TC01: Geçerli Bilgilerle yeni course oluşturmak"""
    print("\n--- TC01 Başlatılıyor ---")
    endpoint = f"{BASE_URL}/api/addCourse"
    
    body = {
        "title": "course görmek güzeldir",
        "type": "course",
        "slug": "course-gormek-guzeldir",
        "start_date": "2025-11-28",
        "duration": 40,
        "capacity": 40,
        "price": 0,
        "description": "Yeni oluşturulan kursun açıklamasıdır.",
        "teacher_id": 1016,
        "category_id": 611
    }
    
    response = requests.post(endpoint, json=body, headers=HEADERS)
    
    assert response.status_code == 200, f"Beklenen 200, Gelen {response.status_code}"
    
    json_data = response.json()
    assert json_data.get("remark") == "success"
    # API bazen "Message" bazen "message" dönebilir, Java kodunda "Message" kullanılmış
    assert json_data.get("Message") == "Successfully Added."
    
    print("TC01 Başarılı.")

def test_post_courses_tc02():
    """US003 TC02: Eksik data (teacher_id) ile POST"""
    print("\n--- TC02 Başlatılıyor ---")
    endpoint = f"{BASE_URL}/api/addCourse"
    
    # teacher_id eksik
    body = {
        "title": "course görmek güzeldir",
        "type": "course",
        "slug": "course-gormek-guzeldir",
        "start_date": "2025-11-28",
        "duration": 40,
        "capacity": 40,
        "price": 0,
        "description": "Yeni oluşturulan kursun açıklamasıdır.",
        # "teacher_id": 1016,  <-- Yorum satırına alındı
        "category_id": 611
    }
    
    response = requests.post(endpoint, json=body, headers=HEADERS)
    
    assert response.status_code == 422, f"Beklenen 422, Gelen {response.status_code}"
    assert response.json().get("message") == "The teacher id field is required."
    
    print("TC02 Başarılı.")

def test_post_courses_tc03():
    """US003 TC03: Geçersiz Token ile POST"""
    print("\n--- TC03 Başlatılıyor ---")
    endpoint = f"{BASE_URL}/api/addCourse"
    
    body = {
        "title": "course deneme",
        "type": "course",
        "teacher_id": 1016,
        "category_id": 611
    }
    
    invalid_headers = HEADERS.copy()
    invalid_headers["Authorization"] = "Bearer invalid"
    
    response = requests.post(endpoint, json=body, headers=invalid_headers)
    
    assert response.status_code == 401
    assert response.json().get("message") == "Unauthenticated."
    
    print("TC03 Başarılı.")

def test_post_courses_tc04():
    """US004: Zincirleme Test (Create -> Extract ID -> Get Verify)"""
    print("\n--- TC04 (Zincirleme Test) Başlatılıyor ---")
    
    # 1. Adım: Kursu Oluştur (POST)
    test_post_courses_tc01() # TC01'deki aynı body ve işlemi kullanabiliriz veya tekrar yazabiliriz.
    # Burada mantığı açıkça görelim diye tekrar yazıyorum ama normalde fonksiyon çağrılabilir.
    
    endpoint_post = f"{BASE_URL}/api/addCourse"
    body = {
        "title": "course görmek güzeldir",
        "type": "course",
        "slug": "course-gormek-guzeldir",
        "start_date": "2025-11-28",
        "duration": 40,
        "capacity": 40,
        "price": 0,
        "description": "Yeni oluşturulan kursun açıklamasıdır.",
        "teacher_id": 1016,
        "category_id": 611
    }
    response_post = requests.post(endpoint_post, json=body, headers=HEADERS)
    assert response_post.status_code == 200
    
    # 2. Adım: ID'yi al
    # Python'da boşluklu key'leri almak için direkt string olarak yazarız
    new_course_id = response_post.json().get("Added Course ID")
    print(f"Yeni Oluşturulan Kurs ID: {new_course_id}")
    
    # 3. Adım: GET ile doğrula
    endpoint_get = f"{BASE_URL}/api/course/{new_course_id}"
    response_get = requests.get(endpoint_get, headers=HEADERS)
    
    assert response_get.status_code == 200
    
    data = response_get.json().get("data", {})
    assert data.get("id") == new_course_id
    
    # Nested yapı kontrolü: data.translations[0].title
    translations = data.get("translations", [])
    if translations:
        assert translations[0].get("title") == "course görmek güzeldir"
    else:
        raise AssertionError("Translations listesi boş döndü!")
        
    print("TC04 Başarılı: Kayıt oluşturuldu ve doğrulandı.")

if __name__ == "__main__":
    try:
        test_post_courses_tc01()
        test_post_courses_tc02()
        test_post_courses_tc03()
        test_post_courses_tc04()
        print("\n=== US003 TÜM TESTLER BAŞARILI ===")
    except AssertionError as e:
        print(f"\n!!! TEST HATASI !!!\n{e}")
    except Exception as e:
        print(f"\n!!! BEKLENMEYEN HATA !!!\n{e}")

File: src/test/java/Api/TestNg/Courses/US003PostCourses.java
package Api.TestNg.Courses;

import Api.Utilities.API_Methods;
import Api.Utilities.HooksAPI;
import org.testng.annotations.Test;

import java.util.HashMap;

public class US003PostCourses {

    @Test
    public void postCoursesTC01() {

        //   @US03_TC01
        // admin(true) id(null)
        //  Scenario: US03_TC01 Geçerli Bilgilerle yeni course olusturmak
        //    Given api kullanicisi "admin" token ile base url olusturur.
        //    And api kullanicisi "api/addCourse" path parametrelerini olusturur.
        //    Then api kullanicisi addCourse icin request body hazırlar
        //    And api kullanicisi "POST" request gonderir ve donen response'i kaydeder.
        //    Then api kullanicisi status code'un 200 oldugunu dogrular.
        //    And api kullanicisi response body'deki "remark" bilgisinin "success" oldugunu dogrular.
        //    Then api kullanicisi response body'deki "Message" bilgisinin "Successfully Added." oldugunu dogrular.

        // /api/addCourse endpoint'ine gecerli authorization bilgileri ve
        // dogru datalar (title, type, slug, start_date, duration, capacity, price,
        // description, teacher_id) iceren bir POST body gönderildiginde
        // dönen status code'in 200, response body'deki remark bilgisinin "success"
        // ve Message bilgisinin de "Successfully Added." oldugu dogrulanmali.

        //title, type, slug, start_date, duration, capacity, price, description, teacher_id

        HooksAPI.setUpApi("admin");

        API_Methods.pathParam("api/addCourse");

        HashMap<String, Object> requestBody = new HashMap<>();
        requestBody.put("title", "course görmek güzeldir");
        requestBody.put("type", "course");
        requestBody.put("slug", "course-gormek-guzeldir");
        requestBody.put("start_date", "2025-11-28");
        requestBody.put("duration", 40);
        requestBody.put("capacity", 40);
        requestBody.put("price", 0);
        requestBody.put("description", "Yeni oluşturulan kursun açıklamasıdır.");
        requestBody.put("teacher_id", 1016);
        requestBody.put("category_id", 611);

        API_Methods.sendRequest("POST", requestBody);

        API_Methods.statusCodeAssert(200);
        API_Methods.assertBody("remark", "success");
        API_Methods.assertBody("Message", "Successfully Added.");
    }


    @Test
    public void postCoursesTC02() {

        // US003 TC02
        // admin(true) id(null)


        // /api/addCourse endpoint'ine gecerli authorization bilgileri
        // ve eksik data (teacher_id) iceren bir
        // POST body (title, type, slug, start_date, duration, capacity, price, description)
        // gönderildiginde dönen status code'in 422 oldugu ve
        // response body'deki message bilgisinin "The teacher id field is required."
        // oldugu, data icermeyen bir POST request gönderildiginde de dönen status
        // code'in 422 oldugu ve response body'deki message bilgisinin
        // "The teacher id field is required. (and 2 more errors)" oldugu dogrulanmali

        HooksAPI.setUpApi("admin");

        API_Methods.pathParam("api/addCourse");

        HashMap<String, Object> requestBody = new HashMap<>();
        requestBody.put("title", "course görmek güzeldir");
        requestBody.put("type", "course");
        requestBody.put("slug", "course-gormek-guzeldir");
        requestBody.put("start_date", "2025-11-28");
        requestBody.put("duration", 40);
        requestBody.put("capacity", 40);
        requestBody.put("price", 0);
        requestBody.put("description", "Yeni oluşturulan kursun açıklamasıdır.");
        //requestBody.put("teacher_id", 1016);
        requestBody.put("category_id", 611);

        API_Methods.sendRequest("POST", requestBody);

        API_Methods.statusCodeAssert(422);
        API_Methods.assertBody("message", "The teacher id field is required.");


    }

    @Test
    public void postCoursesTC03() {

        // /api/addCourse endpoint'ine gecersiz (invalid token) authorization bilgileri ile
        // dogru datalar (title, type, slug, start_date, duration, capacity, price, description, teacher_id)
        // iceren bir POST body gönderildiginde
        // dönen status code'in 401 oldugu
        // ve response body'deki message bilgisinin "Unauthenticated." oldugu dogrulanmali

        HooksAPI.setUpApi("invalid");

        API_Methods.pathParam("api/addCourse");

        HashMap<String, Object> requestBody = new HashMap<>();
        requestBody.put("title", "course görmek güzeldir");
        requestBody.put("type", "course");
        requestBody.put("slug", "course-gormek-guzeldir");
        requestBody.put("start_date", "2025-11-28");
        requestBody.put("duration", 40);
        requestBody.put("capacity", 40);
        requestBody.put("price", 0);
        requestBody.put("description", "Yeni oluşturulan kursun açıklamasıdır.");
        requestBody.put("teacher_id", 1016);
        requestBody.put("category_id", 611);

        API_Methods.sendRequest("POST", requestBody);

        API_Methods.statusCodeAssert(401);
        API_Methods.assertBody("message", "Unauthenticated.");
    }

    @Test
    public void postCoursesTC04() {

        // US004
        // admin(true) id(null)
        // API uzerinden olusturulmak istenen yeni course kaydinin olustugu API uzerinden dogrulanmali.
        //(Response bodyde dönen Added Course ID ile /api/course/{id}
        // endpoint'ine GET request gönderilerek kayıt oluşturulduğu doğrulanabilir.)

        HooksAPI.setUpApi("admin");

        API_Methods.pathParam("api/addCourse");

        HashMap<String, Object> requestBody = new HashMap<>();
        requestBody.put("title", "course görmek güzeldir");
        requestBody.put("type", "course");
        requestBody.put("slug", "course-gormek-guzeldir");
        requestBody.put("start_date", "2025-11-28");
        requestBody.put("duration", 40);
        requestBody.put("capacity", 40);
        requestBody.put("price", 0);
        requestBody.put("description", "Yeni oluşturulan kursun açıklamasıdır.");
        requestBody.put("teacher_id", 1016);
        requestBody.put("category_id", 611);

        // 1. Adim: Kursu Olustur (POST)
        API_Methods.sendRequest("POST", requestBody);

        // Olusturma isleminin basarili oldugunu dogrula
        API_Methods.statusCodeAssert(200);
        API_Methods.assertBody("remark", "success");
        API_Methods.assertBody("Message", "Successfully Added.");

        // 2. Adim: Olusan Kursun ID'sini Response'dan al (Extract)
        // API donusunde "Added Course ID" anahtari ile ID veriyor
        int newCourseId = API_Methods.response.jsonPath().getInt("'Added Course ID'");
        System.out.println("Yeni Olusturulan Kurs ID: " + newCourseId);

        // 3. Adim: Bu ID ile kaydin olustugunu dogrula (GET)
        // Endpoint: api/course/{id}
        API_Methods.pathParam("api/course/" + newCourseId);

        API_Methods.sendRequest("GET", null);

        // 4. Adim: Get sorgusunun saglamasini yap
        API_Methods.statusCodeAssert(200);
        API_Methods.assertBody("data.id", newCourseId);
        API_Methods.assertBody("data.translations[0].title", "course görmek güzeldir");
    }
}
